filepathfilter/filepathfilter.go

```diff
@@ -57,311 +49,129 @@ func patternsToStrings(ps ...Pattern) []string {
-// string "" will be returned in place of the pattern.
-func (f *Filter) AllowsPattern(filename string) (pattern string, allowed bool) {
-	if f == nil {
-		return "", true
+	tracerx.Printf("filepathfilter: rejecting %q via %v", filename, f.include)
+	if !matched && len(f.include) > 0 {
+		return false
 	}
 
-	if len(f.include)+len(f.exclude) == 0 {
-		return "", true
+	for _, ex := range f.exclude {
+		if ex.Match(filename) {
+			tracerx.Printf("filepathfilter: rejecting %q via %q", filename, ex.String())
+			return false
+		}
 	}
 
-	cleanedName := filepath.Clean(filename)
+	tracerx.Printf("filepathfilter: accepting %q", filename)
+	return true
+}
 
-	if len(f.include) > 0 {
-		matched := false
-		for _, inc := range f.include {
-			matched = inc.Match(cleanedName)
-			if matched {
-				pattern = inc.String()
-				break
-			}
-		}
-		if !matched {
-			return "", false
-		}
-	}
+type wm struct {
+	w    *wildmatch.Wildmatch
+	p    string
+	dirs bool
+}
 
-	if len(f.exclude) > 0 {
-		for _, ex := range f.exclude {
-			if ex.Match(cleanedName) {
-				return ex.String(), false
-			}
+func (w *wm) Match(filename string) bool {
+	return w.w.Match(w.chomp(filename))
+}
+
+func (w *wm) chomp(filename string) string {
+	for _, suffix := range []string{`/`, `\`} {
+		if strings.HasSuffix(filename, suffix) {
+			return strings.TrimSuffix(filename, suffix)
 		}
 	}
+	return filename
+}
 
-	return pattern, true
+func (w *wm) String() string {
+	return w.p
 }
 
 const (
-	sep = string(filepath.Separator)
+	sep byte = '/'
 )
 
-func NewPattern(rawpattern string) Pattern {
-	cleanpattern := filepath.Clean(rawpattern)
-
-	// Special case local dir, matches all (inc subpaths)
-	if _, local := localDirSet[cleanpattern]; local {
-		return &noOpMatcher{pattern: cleanpattern}
-	}
-
-	hasPathSep := strings.Contains(cleanpattern, sep)
-	ext := filepath.Ext(cleanpattern)
-	plen := len(cleanpattern)
-	if plen > 1 && !hasPathSep && strings.HasPrefix(cleanpattern, "*") && cleanpattern[1:plen] == ext {
-		return &simpleExtPattern{ext: ext}
-	}
-
-	// special case * when there are no path separators
-	// filepath.Match never allows * to match a path separator, which is correct
-	// for gitignore IF the pattern includes a path separator, but not otherwise
-	// So *.txt should match in any subdir, as should test*, but sub/*.txt would
-	// only match directly in the sub dir
-	// Don't need to test cross-platform separators as both cleaned above
-	if !hasPathSep && strings.Contains(cleanpattern, "*") {
-		pattern := regexp.QuoteMeta(cleanpattern)
-		regpattern := fmt.Sprintf("^%s$", strings.Replace(pattern, "\\*", ".*", -1))
-		return &pathlessWildcardPattern{
-			rawPattern: cleanpattern,
-			wildcardRE: regexp.MustCompile(regpattern),
+func NewPattern(p string) Pattern {
+	pp := p
+
+	// Special case: the below patterns match anything according to existing
+	// behavior.
+	switch pp {
+	case `*`, `*.*`, `.`, `./`, `.\`:
```

Comment:
`.` matched all files? How did this work? Git would clean/smudge all files with `.`, right? Wouldn't that only affect the Git LFS *only* include/exclude patterns?

Reply:
Indeed, `.` used to match all files [[1]](https://github.com/git-lfs/git-lfs/blob/v2.3.4/filepathfilter/filepathfilter.go#L358). This is meant to preserve behavior between versions.